src/Symfony/Component/DependencyInjection/Compiler/ServiceFactoryTagPass.php

```diff
@@ -0,0 +1,115 @@
+    {
+        if (!$value instanceof Definition || !$value->hasTag('container.service_factory')) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        if (!$r = $this->container->getReflectionClass($class = $value->getClass())) {
+            throw new InvalidArgumentException(sprintf('Class "%s" used for service factory "%s" cannot be found.', $class, $this->currentId));
+        }
+
+        if ($r->isAbstract()) {
+            throw new InvalidArgumentException(sprintf('Class "%s" used for service factory "%s" cannot be abstract.', $class, $this->currentId));
+        }
+
+        foreach ($this->getMethodsToFactorize($r) as $id => $method) {
+            $this->container->register($id)
+                ->setFactory(array(new Reference($this->currentId), $method->getName()))
+                ->setClass($this->getClass($method));
+        }
+    }
+
+    /**
+     * Gets the list of methods to factorize.
+     *
+     * @param \ReflectionClass $reflectionClass
+     *
+     * @return \ReflectionMethod[]
+     */
+    private function getMethodsToFactorize(\ReflectionClass $reflectionClass)
+    {
+        $methodsToFactorize = array();
+
+        foreach ($reflectionClass->getMethods() as $method) {
+            if ($method->isConstructor()) {
+                continue;
+            }
+
+            while (true) {
+                if (false !== $doc = $method->getDocComment()) {
+                    if (false !== stripos($doc, '@service') && preg_match('#(?:^/\*\*|\n\s*+\*)\s*+@service(?:\s(\S+)?|\*/$)#i', $doc, $matches)) {
+                        if (!$method->isPublic()) {
+                            throw new InvalidArgumentException(sprintf('Method "%s::%s()" must be public in order to create a service from it.', $reflectionClass->getName(), $name));
+                        }
+
+                        $id = isset($matches[1]) ? $matches[1] : Container::underscore($method->getName());
+                        if (isset($methodsToFactorize[$id]) || $this->container->has($id)) {
+                            throw new InvalidArgumentException(sprintf('Cannot create service "%s" from service factory "%s" as it already exists.', $id, $this->currentId));
+                        }
+
+                        $methodsToFactorize[$id] = $method;
+                        break;
+                    }
+                    if (false === stripos($doc, '@inheritdoc') || !preg_match('#(?:^/\*\*|\n\s*+\*)\s*+(?:\{@inheritdoc\}|@inheritdoc)(?:\s|\*/$)#i', $doc)) {
+                        break;
+                    }
+                }
+                try {
+                    $method = $method->getPrototype();
+                } catch (\ReflectionException $e) {
+                    break; // method has no prototype
+                }
+            }
+        }
+
+        return $methodsToFactorize;
+    }
+
+    private function getClass(\ReflectionMethod $reflectionMethod)
```

Comment:
you should be able to use `Symfony\Component\DependencyInjection\LazyProxy\ProxyHelper::getTypeHint()` instead